### PR TITLE
[IDP] use SSO email if user is org-owned user

### PR DESCRIPTION
**Pull request by iQQBot**
## Description

[IDP] use SSO email if user is org-owned user

## Related Issue(s)

Fixes #

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - pd-idp-use-sso-email</li>
	<li><b>🔗 URL</b> - <a href="https://pd-idp-use-sso-email.preview.gitpod-dev.com/workspaces" target="_blank">pd-idp-use-sso-email.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [x] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
</details>

/hold

### CODE REVIEW

File: components/gitpod-protocol/go/gitpod-service.go
Comment (easyCZ) [inline components/gitpod-protocol/go/gitpod-service.go:2164]: Can we make this a `time.Time` on the protocol level? The RPC should then convert appropriately when it either sends the request, or receives it.

File: components/gitpod-protocol/go/gitpod-service.go
Comment (easyCZ) [inline components/gitpod-protocol/go/gitpod-service.go:2164]: Does this actually work without transforming it to a string? Server uses ISO 8601 as the date format, while go will serialize into RFC 3339

File: components/public-api-server/pkg/apiv1/identityprovider.go
Comment (easyCZ) [inline components/public-api-server/pkg/apiv1/identityprovider.go:84]: For readability, I recommend pulling this into a separate function where you return the moment you find the first valid set. The use of the `email == "" || ...` as a continuation is confusing, and non-obvious when scanning the code

File: components/gitpod-protocol/go/gitpod-service.go
Comment (easyCZ) [reply components/gitpod-protocol/go/gitpod-service.go:2164]: `2020-02-08 09+07:00` is valid ISO 8601 but not valid RFC 3339

Comment (easyCZ) [reply components/gitpod-protocol/go/gitpod-service.go:2164]: Yes, not everything is easy.

**Review by easyCZ — APPROVED** [review]

**Comment by easyCZ** [comment]
Approving, but with reservations about date format mismatch.

**Comment by easyCZ** [comment]
FYI, a proper solution to this in more general is outlined in https://gitpod.slack.com/archives/C02EN94AEPL/p1686566418075449. While the duplication of the server logic over to Public API ensures "parity", it is only a bandage on the bleed

**Comment by loujaybee** [comment]
> A proper solution to this in more general is outlined in - @easyCZ 

Is it worth waiting for some of those changes before we merge this PR? 

Or, we'd be okay to do them as a follow up?

File: components/gitpod-protocol/go/gitpod-service.go
Comment (loujaybee) [inline components/gitpod-protocol/go/gitpod-service.go:2164]: **Question:** Is `lastSigninTime` the right property for us to rely on? That feels quite implicit. Should we attach some other more descriptive property to the user object, such as being "org-owned" user, etc? 🤔

Comment (AlexTugarev) [reply components/gitpod-protocol/go/gitpod-service.go:2164]: Yes it is implicit.
Also, the user is to be marked as org-owned, which is already happening. If `User.organizationId` is set, that account is considered org-owned.

Indeed, it would be great to persist some of those details explicitly. Thinking out loud about the shape: 

```
interface User {
  // ...
  emails: {
    email: string;
    updatedAt: time;
    provider?: string; // github.com, idp.comp.org
    verified: boolean;
    organizationId: uuid;
    personal: boolean;
    // ...
  }[]
}
```

**Pull request closed without merging**
